File: intelitor-v5.2/lib/indrajaal/observability/fractal/hlc.ex
defmodule Indrajaal.Observability.Fractal.HLC do
  @moduledoc """
  Hybrid Logical Clock (HLC) for Fractal Logging System.

  Provides causal ordering of events without tight NTP synchronization.
  L3+ logs MUST use HLC timestamps per SC-LOG-006.

  ## Format

  HLC = {Physical, Counter, NodeID}

  - Physical: Unix microseconds (48 bits usable)
  - Counter: Logical counter (16 bits, 0-65_535)
  - NodeID: Unique node identifier

  ## Usage

      hlc = Indrajaal.Observability.Fractal.HLC.now()
      # => %{physical: 1_735_123_456_789_000, counter: 0, node_id: "node-abc123"}

  ## STAMP Compliance

  - SC-LOG-006: L3+ logs MUST use HLC timestamps
  """

  @type timestamp :: %{
          physical: non_neg_integer(),
          counter: non_neg_integer(),
          node_id: String.t()
        }

  @max_counter 65_535
  @max_drift_ms 50

  # ============================================================
  # STATE (Agent-based for thread-safe counter)
  # ============================================================

  use Agent
  require Logger

  @doc false
  def start_link(_opts \\ []) do
    Agent.start_link(
      fn -> %{last_physical: 0, counter: 0, node_id: generate_node_id()} end,
      name: __MODULE__
    )
  end

  # ============================================================
  # PUBLIC API
  # ============================================================

  @doc """
  Generate a new HLC timestamp.

  Guarantees:
  - Monotonically increasing
  - Causal ordering preserved
  - Max drift of #{@max_drift_ms}ms from wall clock
  """
  @spec now() :: timestamp()
  def now do
    ensure_started()

    Agent.get_and_update(__MODULE__, fn state ->
      physical = System.system_time(:microsecond)
      {new_state, hlc} = advance(state, physical)
      {hlc, new_state}
    end)
  end

  @doc """
  Update the local HLC based on a received remote timestamp.

  Used for distributed systems to maintain causality.
  """
  @spec update(timestamp()) :: timestamp()
  def update(remote) do
    ensure_started()

    Agent.get_and_update(__MODULE__, fn state ->
      physical = System.system_time(:microsecond)
      {new_state, hlc} = merge(state, remote, physical)
      {hlc, new_state}
    end)
  end

  @doc """
  Compare two HLC timestamps.

  Returns:
  - `:lt` if a < b
  - `:eq` if a == b
  - `:gt` if a > b
  """
  @spec compare(timestamp(), timestamp()) :: :lt | :eq | :gt
  def compare(a, b) do
    cond do
      a.physical < b.physical -> :lt
      a.physical > b.physical -> :gt
      a.counter < b.counter -> :lt
      a.counter > b.counter -> :gt
      true -> :eq
    end
  end

  @doc """
  Encode HLC to a compact binary format (12 bytes).
  """
  @spec encode(timestamp()) :: binary()
  def encode(hlc) do
    <<hlc.physical::unsigned-integer-size(64), hlc.counter::unsigned-integer-size(16)>>
  end

  @doc """
  Decode HLC from binary format.
  """
  @spec decode(binary()) :: {:ok, timestamp()} | {:error, :invalid_format}
  def decode(<<physical::unsigned-integer-size(64), counter::unsigned-integer-size(16)>>) do
    {:ok, %{physical: physical, counter: counter, node_id: ""}}
  end

  def decode(_), do: {:error, :invalid_format}

  @doc """
  Get the current node ID.
  """
  @spec node_id() :: String.t()
  def node_id do
    ensure_started()
    Agent.get(__MODULE__, fn state -> state.node_id end)
  end

  # ============================================================
  # PRIVATE: CLOCK ADVANCEMENT
  # ============================================================

  defp advance(state, physical) do
    # Physical time moved forward
    if physical > state.last_physical do
      hlc = %{physical: physical, counter: 0, node_id: state.node_id}
      new_state = %{state | last_physical: physical, counter: 0}
      {new_state, hlc}
    else
      # Same or backwards - increment counter
      new_counter = min(state.counter + 1, @max_counter)

      if new_counter >= @max_counter do
        # Counter overflow - wait for physical time to advance
        :timer.sleep(1)
        advance(state, System.system_time(:microsecond))
      else
        hlc = %{physical: state.last_physical, counter: new_counter, node_id: state.node_id}
        new_state = %{state | counter: new_counter}
        {new_state, hlc}
      end
    end
  end

  defp merge(state, remote, physical) do
    max_physical = Enum.max([physical, state.last_physical, remote.physical])

    # Check for excessive drift
    drift_ms = div(max_physical - physical, 1000)

    if drift_ms > @max_drift_ms do
      # Clock skew detected - use local time with warning
      Logger.warning("[Fractal.HLC] Clock skew detected: #{drift_ms}ms")
      advance(state, physical)
    else
      new_counter =
        cond do
          max_physical == state.last_physical and max_physical == remote.physical ->
            Enum.max([state.counter, remote.counter]) + 1

          max_physical == state.last_physical ->
            state.counter + 1

          max_physical == remote.physical ->
            remote.counter + 1

          true ->
            0
        end

      hlc = %{
        physical: max_physical,
        counter: min(new_counter, @max_counter),
        node_id: state.node_id
      }

      new_state = %{state | last_physical: max_physical, counter: new_counter}
      {new_state, hlc}
    end
  end

  # ============================================================
  # PRIVATE: INITIALIZATION
  # ============================================================

  defp ensure_started do
    case Process.whereis(__MODULE__) do
      nil -> start_link()
      _pid -> :ok
    end
  end

  defp generate_node_id do
    node_str = node() |> to_string()
    suffix_bytes = :crypto.strong_rand_bytes(3)
    random_suffix = suffix_bytes |> Base.encode16(case: :lower)
    "#{node_str}-#{random_suffix}"
  end
end

File: intelitor-v5.2/lib/cepaf/src/Cepaf/Observability/Fractal/HLC.fs
namespace Cepaf.Observability.Fractal

open System
open System.Threading

/// Hybrid Logical Clock for causal ordering without NTP synchronization.
/// Based on Zenoh protocol principles.
/// STAMP Compliance: SC-LOG-006 (L3+ logs MUST use HLC timestamps)
module HLC =

    // ============================================================
    // TYPES
    // ============================================================

    /// HLC Timestamp with physical time, logical counter, and node ID
    [<Struct>]
    type Timestamp = {
        /// Physical time in microseconds (Unix epoch)
        Physical: int64

        /// Logical counter for events at same physical time (0-65535)
        Counter: int

        /// Node identifier (machine name or UUID)
        NodeId: string
    }

    /// HLC state for a node
    type State = {
        mutable Physical: int64
        mutable Counter: int
        NodeId: string
        Lock: obj
    }

    // ============================================================
    // GLOBAL STATE
    // ============================================================

    let private state: State = {
        Physical = 0L
        Counter = 0
        NodeId = Environment.MachineName
        Lock = obj()
    }

    // ============================================================
    // CORE OPERATIONS
    // ============================================================

    /// Get current physical time in microseconds
    let private getPhysicalTime () : int64 =
        DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() * 1000L

    /// Generate a new HLC timestamp for a local event
    let now () : Timestamp =
        lock state.Lock (fun () ->
            let physical = getPhysicalTime()

            if physical > state.Physical then
                // Physical time has advanced
                state.Physical <- physical
                state.Counter <- 0
            else
                // Same or earlier physical time, increment counter
                state.Counter <- state.Counter + 1

            {
                Physical = state.Physical
                Counter = state.Counter
                NodeId = state.NodeId
            }
        )

    /// Update HLC from a received timestamp (message receive)
    let update (received: Timestamp) : Timestamp =
        lock state.Lock (fun () ->
            let physical = getPhysicalTime()

            if physical > state.Physical && physical > received.Physical then
                // Local physical time is most recent
                state.Physical <- physical
                state.Counter <- 0
            elif received.Physical > state.Physical then
                // Received physical time is most recent
                state.Physical <- received.Physical
                state.Counter <- received.Counter + 1
            elif received.Physical = state.Physical then
                // Same physical time, take max counter + 1
                state.Counter <- max state.Counter received.Counter + 1
            else
                // Local is ahead, increment counter
                state.Counter <- state.Counter + 1

            {
                Physical = state.Physical
                Counter = state.Counter
                NodeId = state.NodeId
            }
        )

    /// Merge two timestamps (for causality tracking)
    let merge (a: Timestamp) (b: Timestamp) : Timestamp =
        if a.Physical > b.Physical then
            { a with Counter = a.Counter + 1 }
        elif b.Physical > a.Physical then
            { b with Counter = b.Counter + 1 }
        else
            {
                Physical = a.Physical
                Counter = max a.Counter b.Counter + 1
                NodeId = state.NodeId
            }

    // ============================================================
    // COMPARISON
    // ============================================================

    /// Compare two HLC timestamps
    /// Returns: -1 if a < b, 0 if a = b, 1 if a > b
    let compare (a: Timestamp) (b: Timestamp) : int =
        match compare a.Physical b.Physical with
        | 0 ->
            // Same physical time, compare counters
            match compare a.Counter b.Counter with
            | 0 ->
                // Same counter, compare node IDs (tie-breaker)
                // Normalize String.Compare result to -1/0/1
                let cmp = String.Compare(a.NodeId, b.NodeId, StringComparison.Ordinal)
                if cmp < 0 then -1 elif cmp > 0 then 1 else 0
            | c -> c
        | c -> c

    /// Check if a happened before b
    let happenedBefore (a: Timestamp) (b: Timestamp) : bool =
        compare a b < 0

    /// Check if a happened after b
    let happenedAfter (a: Timestamp) (b: Timestamp) : bool =
        compare a b > 0

    /// Check if timestamps are concurrent (neither happened before the other)
    let areConcurrent (a: Timestamp) (b: Timestamp) : bool =
        // Concurrent if same physical time but different nodes
        a.Physical = b.Physical && a.NodeId <> b.NodeId

    // ============================================================
    // ARITHMETIC
    // ============================================================

    /// Add milliseconds to a timestamp
    let addMs (ts: Timestamp) (ms: int64) : Timestamp =
        { ts with Physical = ts.Physical + (ms * 1000L); Counter = 0 }

    /// Subtract two timestamps to get duration in microseconds
    let diffMicros (a: Timestamp) (b: Timestamp) : int64 =
        a.Physical - b.Physical

    /// Subtract two timestamps to get duration in milliseconds
    let diffMs (a: Timestamp) (b: Timestamp) : int64 =
        (a.Physical - b.Physical) / 1000L

    // ============================================================
    // SERIALIZATION
    // ============================================================

    /// Serialize to compact binary format (12 bytes)
    /// Format: Physical (8 bytes) + Counter (2 bytes) + NodeHash (2 bytes)
    let toBytes (ts: Timestamp) : byte[] =
        let bytes = Array.zeroCreate<byte> 12
        BitConverter.GetBytes(ts.Physical) |> Array.iteri (fun i b -> bytes.[i] <- b)
        BitConverter.GetBytes(uint16 ts.Counter) |> Array.iteri (fun i b -> bytes.[8 + i] <- b)
        let nodeHash = uint16 (abs (ts.NodeId.GetHashCode()) % 65536)
        BitConverter.GetBytes(nodeHash) |> Array.iteri (fun i b -> bytes.[10 + i] <- b)
        bytes

    /// Deserialize from compact binary format
    let fromBytes (bytes: byte[]) (nodeId: string option) : Result<Timestamp, string> =
        if bytes.Length < 12 then
            Error "Invalid HLC bytes: expected 12 bytes"
        else
            try
                let physical = BitConverter.ToInt64(bytes, 0)
                let counter = int (BitConverter.ToUInt16(bytes, 8))
                Ok {
                    Physical = physical
                    Counter = counter
                    NodeId = defaultArg nodeId "unknown"
                }
            with ex ->
                Error $"Failed to parse HLC bytes: {ex.Message}"

    /// Convert to ISO 8601 string with counter
    let toIso8601 (ts: Timestamp) : string =
        let dt = DateTimeOffset.FromUnixTimeMilliseconds(ts.Physical / 1000L)
        $"{dt:o}.{ts.Counter:D4}"

    /// Convert to human-readable string
    let toString (ts: Timestamp) : string =
        $"{ts.Physical}.{ts.Counter}@{ts.NodeId}"

    /// Parse from string representation
    let parse (s: string) : Result<Timestamp, string> =
        try
            let parts = s.Split('@')
            let timeParts = parts.[0].Split('.')

            Ok {
                Physical = Int64.Parse(timeParts.[0])
                Counter = if timeParts.Length > 1 then Int32.Parse(timeParts.[1]) else 0
                NodeId = if parts.Length > 1 then parts.[1] else state.NodeId
            }
        with ex ->
            Error $"Failed to parse HLC string '{s}': {ex.Message}"

    // ============================================================
    // DIAGNOSTICS
    // ============================================================

    /// Get current HLC state for diagnostics
    let getState () =
        {|
            Physical = state.Physical
            Counter = state.Counter
            NodeId = state.NodeId
            CurrentTime = getPhysicalTime()
            Drift = getPhysicalTime() - state.Physical
        |}

    /// Reset HLC state (for testing only)
    let reset () =
        lock state.Lock (fun () ->
            state.Physical <- 0L
            state.Counter <- 0
        )

    // ============================================================
    // VALIDATION (SC-LOG-006)
    // ============================================================

    /// Validate that a timestamp is valid for a fractal level
    let validateForLevel (level: FractalLevel) (ts: Timestamp option) : Result<unit, string> =
        match level with
        | FractalLevel.L3 | FractalLevel.L4 | FractalLevel.L5 ->
            match ts with
            | Some t when t.Physical > 0L -> Ok ()
            | Some _ -> Error "HLC timestamp has invalid physical time"
            | None -> Error $"SC-LOG-006: {FractalLevel.toString level} logs MUST have HLC timestamp"
        | _ -> Ok ()  // L1/L2 don't require HLC

    /// Check if timestamp is within acceptable drift
    let isWithinDrift (ts: Timestamp) (maxDriftMs: int64) : bool =
        let now = getPhysicalTime()
        let drift = abs (now - ts.Physical)
        drift <= (maxDriftMs * 1000L)
